### PR TITLE
Allow default route to not be present

### PR DESCRIPTION
**Pull request by haircommander**
First, attempt to find an IP address from /proc/net/route. If we fail to do so, attempt to find a clear primary

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
LGTM

**Comment by sjenning** [comment]
This doesn't seem to work
```
# ip r
10.42.15.0/24 dev ens3 proto kernel scope link src 10.42.15.10 metric 100 
10.88.0.0/16 dev cni0 proto kernel scope link src 10.88.0.1 linkdown 
169.254.169.254 via 10.42.15.2 dev ens3 proto dhcp metric 100 

# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc fq_codel state UP group default qlen 1000
    link/ether fa:16:3e:af:f4:19 brd ff:ff:ff:ff:ff:ff
    inet 10.42.15.10/24 brd 10.42.15.255 scope global dynamic noprefixroute ens3
       valid_lft 86197sec preferred_lft 86197sec
    inet6 fe80::f816:3eff:feaf:f419/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
3: cni0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether ea:25:e4:25:ec:bf brd ff:ff:ff:ff:ff:ff
    inet 10.88.0.1/16 brd 10.88.255.255 scope global cni0
       valid_lft forever preferred_lft forever
    inet6 fe80::e825:e4ff:fe25:ecbf/64 scope link 
       valid_lft forever preferred_lft forever

# systemctl start crio
Job for crio.service failed because the control process exited with error code.
See "systemctl status crio.service" and "journalctl -xe" for details.

# journalctl -u crio.service --since="20 seconds ago"
systemd[1]: Starting Open Container Initiative Daemon...
crio[6470]: time="2019-07-29 21:58:08.860059205Z" level=fatal msg="Failed to find one IP address after failing to choose bind address: no default routes found in "/proc/net/route" or "/proc/net/ipv6_route""
systemd[1]: crio.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: crio.service: Failed with result 'exit-code'.
systemd[1]: Failed to start Open Container Initiative Daemon.
```

File: server/server.go [hunk 1]
```diff
@@ -454,6 +455,55 @@ func New(
 	return s, nil
 }
 
+// getHostIP first attempts to read the IP address of the host from /proc/net/route
+// If that fails (there is no /proc/net/route), it then attempts to deterministically find a default IP
+// It does so by finding exactly one IP address that is not a loop back, and not created by CNI.
+func (s *Server) getHostIP() (net.IP, error) {
```
Comment (sjenning) [inline server/server.go]: we need to make this unit testable

File: server/server.go [hunk 2]
```diff
@@ -454,6 +455,55 @@ func New(
...
+	// First, attempt to find a primary IP from /proc/net/route deterministically
+	hostIP, err := knet.ChooseBindAddress(nil)
+	if err != nil {
```
Comment (sjenning) [inline server/server.go]: we can lose a level of indentation by just 
```
if err == nil {
  return hostIP, nil
}
```

File: server/server.go [hunk 3]
```diff
@@ -454,6 +455,55 @@ func New(
...
+		// if that fails, check if we can find a primary IP address unambiguously
+		ifaces, err2 := net.Interfaces()
```
Comment (sjenning) [inline server/server.go]: then we can just reuse `err2`

File: server/server.go [hunk 4]
```diff
@@ -454,6 +455,55 @@ func New(
...
+		if err2 != nil {
+			return nil, errors.Wrapf(err, "Failed to read network Interfaces after failing to choose bind address")
+		}
+
+		// we are hoping for exactly 1 possible IP
+		numPossibleIPs := 0
+		for _, iface := range ifaces {
```
Comment (sjenning) [inline server/server.go]: could we break this loop into a different function that we can unit test?

Comment (sjenning) [reply server/server.go]: gah, seems like it make a ton of different calls though that we'd have to mock up :disappointed:

**Comment by haircommander** [comment]
@sjenning thanks for the review, I'll get to the unit tests tomorrow :+1: :smile:

**Comment by haircommander** [comment]
@sjenning after some deliberation, i have deemed the function un-unit-testable without having control over the networking stack of the vm setup. I was hoping it could be mocked, or members could be spoofed, but the implementation needs both function implementations and member variables, so no dice :\

Also on @mrunalp's slack messages about using the CNI config to parse the network names, unfortunately ocicni doesn't currently expose that over the CNIPlugin interface, as such I added a comment to mark future work will be needed if the format changes. the cni0 you found is probably podman's cni, as CRI-O wouldn't have set one up by that point. I also verified the cni configurations on a cluster are in the cni* format, which they are.

All in all, I think this is ready, PTAL

**Comment by sjenning** [comment]
LGTM.  want @mrunalp to give final lgtm.

**Comment by rhatdan** [comment]
@mrunalp Looks like this is ready to merge, once you give LGTM.

**Comment by haircommander** [comment]
/hold
I don't think I'm happy with this solution...

**Comment by haircommander** [comment]
/hold cancel

Ok, I came to a solution I am happy with. Here, we are emulating the way the kubelet chooses its nodeIP, with the one change of prioritizing the legacy CRI-O-way of doing it (ChooseHostIP is called earlier than kubelet, which should be fine because the default route is a good choice in general). 

filtering by cni* was both hacky, and not a good long term solution (https://github.com/containers/libpod/pull/3755 was a PR that originally could have broken this assumption). Instead, let's rely on the ordering provided by InterfaceAddrs(), which is more reliable. 

@sjenning, PTAL; can you confirm this works with the setup you used above?

**Comment by cgwalters** [comment]
This is related to https://github.com/cri-o/cri-o/issues/2478 right?

**Comment by sjenning** [comment]
@haircommander should we expect e2e to pass now?

**Comment by haircommander** [comment]
@sjenning nope, we had 3 days of green CI and we're broken again :\ 
@cgwalters sort of, we still depend on routing, but it allows some aspects of routing to not be configured, so this may help your usecase

**Pull request merged**